byterun/caml/misc.h

```diff
@@ -124,6 +124,48 @@ CAMLnoreturn_end;
 CAMLextern char * caml_strdup(const char * s);
 CAMLextern char * caml_strconcat(int n, ...); /* n args of const char * type */
 
+/* Integer arithmetic with overflow detection.
+   The functions return 0 if no overflow, 1 if overflow.
+   The result of the operation is always stored at [*res].
+   If no overflow is reported, this is the exact result.
+   If overflow is reported, this is the exact result modulo 2 to the word size.
+*/
+
+#ifndef __has_builtin
+#define __has_builtin(x) 0
+#endif  
+
+static inline int caml_addu_overflow(uintnat a, uintnat b, uintnat * res)
+{
+#if __GNUC__ >= 5 || __has_builtin(__builtin_add_overflow)
+  return __builtin_add_overflow(a, b, res);
+#else
+  uintnat c = a + b;
+  *res = c;
+  return c < a;
+#endif
+}
+  
+static inline int caml_subu_overflow(uintnat a, uintnat b, uintnat * res)
+{
+#if __GNUC__ >= 5 || __has_builtin(__builtin_sub_overflow)
+  return __builtin_sub_overflow(a, b, res);
+#else
+  uintnat c = a - b;
+  *res = c;
+  return a < b;
+#endif
+}
+  
+#if __GNUC__ >= 5 || __has_builtin(__builtin_mul_overflow)
+static inline int caml_mulu_overflow(uintnat a, uintnat b, uintnat * res)
+{
+  return __builtin_mul_overflow(a, b, res);
+}
+#else
+extern int caml_mulu_overflow(uintnat a, uintnat b, uintnat * res);
```

Comment:
Perhaps the implementation could be put right in the header, with `INLINE`, to help the less smarter compilers/linkers? The function seems perfect for inlining.